vim-patch:8.2.{4242,4315}: put in Visual mode cannot be repeated

Shougo opened the pull request:
Close #1790

Problem:    Put in Visual mode cannot be repeated.
Solution:   Use "P" to put without yanking the deleted text into the unnamed
            register. (Shougo Matsushita, closes vim/vim#9591)
https://github.com/vim/vim/commit/fb55207ed17918c8a2a6cadf5ad9d5fcf686a7ab

continue of https://github.com/neovim/neovim/pull/17097

Shougo committed:
vim-patch:8.2.4242: put in Visual mode cannot be repeated

Problem:    Put in Visual mode cannot be repeated.
Solution:   Use "P" to put without yanking the deleted text into the unnamed
            register. (Shougo Matsushita, closes vim/vim#9591)
https://github.com/vim/vim/commit/fb55207ed17918c8a2a6cadf5ad9d5fcf686a7ab

Shougo commented:
I have updated the documentation.

Shougo commented:
Oh, conflicted...

> CI / ubuntu-20.04 functionaltest-lua (cc=gcc) (pull_request)

It is just process timeout.

Shougo commented:
Fixed the conflict.

Shougo commented:
Passed.

Shougo commented:
Hi, when the patch is merged?

zeertzjq committed:
vim-patch:partial:f10911e5db16

Update runtime files
https://github.com/vim/vim/commit/f10911e5db16f1fe6ab519c5d091ad0c1df0d063

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -7533,6 +7534,12 @@ static void nv_put_opt(cmdarg_T *cap, bool fix_indent)
 
+        if (save_unnamed) {
+          // Note: Discard the result
+          if (op_reg_set_previous('0')) {
```
This doesn't work if the previous register is not register 0. This is not covered by tests.

Shougo replied:
Well, but how to get the current previous name?
neovim does not have Vim's function.  So it is hard to port the function.

zeertzjq replied:
I'm also still thinking about this.

Shougo commented:
Oh, thank you for the fix.

zeertzjq commented:
I'll add more tests: <https://github.com/vim/vim/pull/9708>.

zeertzjq committed:
vim-patch:8.2.4315: put in Visual mode not fully tested

Problem:    Put in Visual mode not fully tested.
Solution:   Add a few more test cases. (closes vim/vim#9708)
https://github.com/vim/vim/commit/6bf821e8abe1da24e5d0624f032d7eda745756e8

zeertzjq commented:
The version of this branch at <https://github.com/neovim/neovim/commit/36cac7610a471f46371eff86624699a17894f624> now fails the newly added tests:
```
1 FAILED:
Found errors in Test_visual_paste():
function RunTheTest[37]..Test_visual_paste line 46: Expected 'baz' but got 'foo'
function RunTheTest[37]..Test_visual_paste line 50: Expected 'baz' but got 'foo'
function RunTheTest[37]..Test_visual_paste line 52: Expected 'bazooxxfobaz' but got 'bazooxxfofoo'
```

Whereas the latest version of this branch passes them.

zeertzjq commented:
I'm not entirely sure if using an index is the best way to save/restore `y_previous`, or is using a pointer better?

seandewar reviewed (dismissed):
I think the index stuff is fine. While {get/set}_y_previous could be ported (which was introduced to keep y_previous static in https://github.com/vim/vim/commit/c3328169d5566b97a6a6921067017e4369dd7cd6), get_unname_register and Nvim's op_reg_set_previous achieves the same.

zeertzjq commented:
I'll just open another PR: #17346

zeertzjq commented:
Superseded by #17346

Shougo commented:
Thanks.
